Specifications/Events/for_EventStore/given/an_event_store_and_an_execution_context.cs

```diff
@@ -0,0 +1,50 @@
+// Copyright (c) Dolittle. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Linq;
+using System.Threading;
+using Dolittle.SDK.Services;
+using Machine.Specifications;
+using Microsoft.Extensions.Logging;
+using Moq;
+using static Dolittle.Runtime.Events.Contracts.EventStore;
+using Contracts = Dolittle.Runtime.Events.Contracts;
+
+namespace Dolittle.SDK.Events.for_EventStore.given
+{
+    public class an_event_store_and_an_execution_context : an_execution_context
+    {
+        protected static Mock<IPerformMethodCalls> caller;
+        protected static Mock<IEventConverter> converter;
+        protected static Mock<IEventTypes> event_types;
+        protected static IEventStore event_store;
+        protected static Contracts.CommitEventsRequest commit_events_request;
+        protected static UncommittedEvents events_to_convert;
+
+        Establish context = () =>
+        {
+            caller = new Mock<IPerformMethodCalls>();
+            event_types = new Mock<IEventTypes>();
+            converter = new Mock<IEventConverter>();
+
+            event_store = new EventStore(caller.Object, converter.Object, execution_context, event_types.Object, Mock.Of<ILogger>());
+
+            // we don't care here what the converter is doing, just return the correct type at least
+            converter.Setup(_ => _.ToSDK(Moq.It.IsAny<Contracts.CommitEventsResponse>()))
+                .Returns(Moq.It.IsAny<CommitEventsResult>);
+
+            // allows us to inspect the uncommitted events just before they'd get committed and make it
+            // return the correct type and amount at least
+            converter.Setup(_ => _.ToProtobuf(Moq.It.IsAny<UncommittedEvents>()))
+                .Callback<UncommittedEvents>((events) => events_to_convert = events)
+                .Returns((UncommittedEvents @events) => @events.Select(_ => new Contracts.UncommittedEvent { }));
```

Comment:
A few things here:

1. There shouldn't be a need for comments in the code. If it is unclear what the code does - it's better to wrap it into a method with a more descriptive name.
2. The `.Returns(Moq...)` on line 34 doesn't work the way you seem to think. It will return the specific thing that you provide - and frankly I'm a bit surprised that this compiles 😅
3. I think you are making your life a bit harder than it needs to be here with returning actual results and tracking the things it sends in to the converter.

I think the tests would be simplified if you just return a specific object (so `CommitEventsResult` and `UncommittedEvents`) from the converter, and check that the `EventStore` implementation passes them along untouched to the other things it uses.

Reply:
1. The setup on the `converter.ToProtobuf` with the callbacks to access the events felt a bit hacky for me so I think that the comment provides some context to it as I don't think it's self evident what I was doing. The upper comment was more for me in the moment that I just left in so I agree on that one.
2. I'm also a bit surprised how everything seemed to work regarding the `.Returns()` as thing's we're also quite happy with accepting null or even outright wrong types here too so that probably kinda drove me deeper into it.
3. What do you mean by returning a specific object (`CommitEventsResult/UncommittedEvents`)?
As in having the converter not be a mock and working correctly (which means that we'd have to setup the `Call()` to have a whole bunch of logic to return a proper `Contracts.CommittedEvent` from the given `Contracts.UncommittedEvents` like the real `EventStoreClient` would do) or the converter would just return an empty `CommitEventsResult/UncommittedEvents`?